[SwiftUI] Add some observable and settable properties to WebPage

rr-codes opened the pull request:
#### 46b7e98a3b6793e6189acd5d72bc58276fc2ca3e
<pre>
[SwiftUI] Add some observable and settable properties to WebPage
<a href="https://bugs.webkit.org/show_bug.cgi?id=283452">https://bugs.webkit.org/show_bug.cgi?id=283452</a>
<a href="https://rdar.apple.com/140312960">rdar://140312960</a>

Reviewed by Aditya Keerthi.

Bring over some of the properties from WKWebView to WebPage, and translate their KVO-ness to become Observable properties.

* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPage_v0.url):
(WebPage_v0.title):
(WebPage_v0.estimatedProgress):
(WebPage_v0.isLoading):
(WebPage_v0.serverTrust):
(WebPage_v0.hasOnlySecureContent):
(WebPage_v0.themeColor):
(WebPage_v0.mediaType):
(WebPage_v0.customUserAgent):
(WebPage_v0.isInspectable):
(WebPage_v0.load(_:)):
(WebPage_v0.load(_:baseURL:)):
(KeyValueObservations.contents):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPage.swift: Renamed from Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageNavigation.swift.
(URL.aboutBlank):
(WebPageTests.basicNavigation):
(WebPageTests.sequenceOfNavigations):
(WebPageTests.navigationWithFailedProvisionalNavigationEvent):
(WebPageTests.observableProperties):

Canonical link: <a href="https://commits.webkit.org/286925@main">https://commits.webkit.org/286925@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/185d860927192e7224b425ae85f1eabf1ed7dc10

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77545 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/56580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30461 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/82140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/28827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79662 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/65730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4877 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/82140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/28827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80612 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/65730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/66560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/82140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/65730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/24055 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/27152 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/65730 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/24393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83536 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/4925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/68993 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/5081 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/66526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/68255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/10375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12004 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/4872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7687 "Build is in progress. Recent messages:") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/4891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-11-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/6650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -24,13 +24,80 @@
 
 import Foundation
+import Observation
+import SwiftUI // FIXME: Do not import SwiftUI in WebKit proper.
```
Do we have a task about a cross import overlay?

Something you might want to front load is opting in to the new "internal imports by default" mode, which would also require you to qualify these as `public import` instead. That should prevent accidentally exposing frameworks going forward, and will be easier to do from the get go.

https://www.swift.org/blog/using-upcoming-feature-flags/
https://github.com/swiftlang/swift-evolution/blob/main/proposals/0409-access-level-on-imports.md

rr-codes replied:
I will update the PR to reflect the bug tracking the overlay adoption. I will also update the PR to qualify the access level imports. I will enable that default import level feature in a subsequent PR

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -24,13 +24,80 @@
+import Observation
+import SwiftUI // FIXME: Do not import SwiftUI in WebKit proper.
 
 @_spi(Internal)
```
I thought we were also going to just stick with `@_spi(Private)` for now?

rr-codes replied:
oops, I keep mixing up Internal and Private

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -24,13 +24,80 @@
+    public var themeColor: Color? {
+        // The color property is a UIColor/NSColor in WKWebView.
+        _$observationRegistrar.access(self, keyPath: \.themeColor)
+        return backingWebView.themeColor.map(Color.init(_:))
```
Aren't the unlabeled conversion initializer to-be-deprecated[^1][^2]? There are also `init(nsColor:)`/`init(uiColor:)` versions.

[^1]: https://developer.apple.com/documentation/swiftui/color/init(_:)-9rl8m
[^2]: https://developer.apple.com/documentation/swiftui/color/init(_:)-1uwrw

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -24,13 +24,80 @@
+        return backingWebView.themeColor.map(Color.init(_:))
+    }
+
+    public var mediaType: String? {
```
IMO, single line `get` and `set` lines (and even implicit `get`s like in the above) can be written as a single line too.
```swift
public var mediaType: String? {
    get { backingWebView.mediaType }
    set { backingWebView.mediaType = newValue }
}
```

Why do these properties not need to also invoke the registrar for change tracking?

rr-codes replied:
because they are not observable / kvo compliant

axiixc replied:
Hmm, we might need to consider this more. Having a few one off non-observable properties is going to be confusing to adopters. I don’t think there’s anything to stop you from doing something like:
```swift
@Bindable var webPage = self.webPage
TextField("Media Type:", text: $webPage.mediaType)
```
(Ignoring that `Binding<String?>` can’t be passed to a text field, but the point stands).

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -39,16 +106,55 @@ import Foundation
         let (stream, continuation) = AsyncStream.makeStream(of: NavigationEvent.self)
-        self.navigations = Navigations(source: stream)
-        self.backingNavigationDelegate = WKNavigationDelegateAdapter(navigationProgressContinuation: continuation)
+        navigations = Navigations(source: stream)
```
I see this isn't new here, but something to consider is whether we want to have a single value here or if the getter for `navigations` should return a fresh sequence every time. I think this ended up being a `struct`, so you are getting an implicitly `Copyable` value for now, but `AsyncSequence` itself doesn't specify whether a sequence may be consumed multiple times or not. We should figure out our behaviors here and make sure we've got an implementation to match.

rr-codes replied:
I added a FIXME to consider and address this

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -39,16 +106,55 @@ import Foundation
+    private func createObservation<Value, BackingValue>(for keyPath: KeyPath<WebPage_v0, Value>, backedBy backingKeyPath: KeyPath<WKWebView, BackingValue>) -> NSKeyValueObservation {
+        backingWebView.observe(backingKeyPath, options: [.prior, .old, .new]) { [_$observationRegistrar, unowned self] _, change in
+            if change.isPrior {
+              _$observationRegistrar.willSet(self, keyPath: keyPath)
```
Nit: indentation

axiixc commented on Tools/TestWebKitAPI/Tests/WebKit Swift/WebPage.swift:
```diff
@@ -189,6 +190,30 @@ struct WebPageNavigationTests {
+        #expect(!page.hasOnlySecureContent)
+        #expect(page.themeColor == nil)
+
+        // FIXME: Make this test more comprehensive once Observation supports observing a stream of changes to properties.
```
Do you have a bug for this?

rr-codes replied:
fixed; I added a reference to the bug in the comment

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -24,31 +24,133 @@
 #if ENABLE_SWIFTUI
 
 import Foundation
+internal import Observation
```
I think this actually would need to be `public` too, since `WebPage_v0` conforms to `Observable` (via macro expansion). It’s probably moot so long as you also `public import SwiftUI` since SwiftUI itself re-exports Observation though.

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -21,34 +21,136 @@
-    @discardableResult public func load(htmlString: String, baseURL: URL) -> NavigationID? {
-        self.backingWebView.loadHTMLString(htmlString, baseURL: baseURL).map(NavigationID.init(_:))
+    private func createObservation<Value, BackingValue>(for keyPath: KeyPath<WebPage_v0, Value>, backedBy backingKeyPath: KeyPath<WKWebView, BackingValue>) -> NSKeyValueObservation {
+        backingWebView.observe(backingKeyPath, options: [.prior, .old, .new]) { [_$observationRegistrar, unowned self] _, change in
```
Nit: I’d put an explicit `return` here. I don’t like using the implicit return for a complex result like this, even though it technically meets the compiler’s requirements. I actually misread this thinking you were dropping the observation token at first 😅

axiixc commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -21,34 +21,136 @@
+}
+
+extension WebPage_v0 {
+    private struct KeyValueObservations: ~Copyable {
```
I’m curious why you’re using this wrapper type instead of just storing these directly on `WebPage` and removing them in `WebPage.deinit`?

rr-codes committed:
[SwiftUI] Add some observable and settable properties to WebPage
https://bugs.webkit.org/show_bug.cgi?id=283452
rdar://140312960

Reviewed by Aditya Keerthi.

Bring over some of the properties from WKWebView to WebPage, and translate their KVO-ness to become Observable properties.

* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(WebPage_v0.url):
(WebPage_v0.title):
(WebPage_v0.estimatedProgress):
(WebPage_v0.isLoading):
(WebPage_v0.serverTrust):
(WebPage_v0.hasOnlySecureContent):
(WebPage_v0.themeColor):
(WebPage_v0.mediaType):
(WebPage_v0.customUserAgent):
(WebPage_v0.isInspectable):
(WebPage_v0.load(_:)):
(WebPage_v0.load(_:baseURL:)):
(KeyValueObservations.contents):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit Swift/WebPage.swift: Renamed from Tools/TestWebKitAPI/Tests/WebKit Swift/WebPageNavigation.swift.
(URL.aboutBlank):
(WebPageTests.basicNavigation):
(WebPageTests.sequenceOfNavigations):
(WebPageTests.navigationWithFailedProvisionalNavigationEvent):
(WebPageTests.observableProperties):

Canonical link: https://commits.webkit.org/286925@main

dpogue commented on Source/WebKit/UIProcess/API/Swift/WebPage.swift:
```diff
@@ -21,34 +21,136 @@
+    }
+
+    public var title: String {
+        self.access(keyPath: \.url)
```
Is this meant to be `keyPath: \.title`?

rr-codes replied:
It is! Good catch!
